File: web/apps/app/lib/utils/reasoning-stream-blocks.test.ts
import { describe, it, expect } from 'vitest';
import { ReasoningBlockSplitter } from './reasoning-stream-blocks';

/**
 * Behavior lock for splitting an expanded thinking stream into stable blocks.
 *
 * Why it exists: react-markdown reparses everything on every render with no caching, and thinking text
 * can run to tens of thousands of characters. Splitting sealed paragraphs into immutable blocks lets each
 * one be memoized and parsed exactly once, making the total cost O(total input).
 * Two invariants: **a block never changes once emitted** (otherwise memoization breaks and the layout
 * reflows), and **a block boundary never cuts into a code fence** (otherwise a code block is split in
 * half and has to be reassembled once it closes).
 */
describe('ReasoningBlockSplitter', () => {
  /** Feed in fixed-size chunks to simulate real chunk arrival */
  function feed(text: string, chunkSize: number) {
    const splitter = new ReasoningBlockSplitter();
    let last = { blocks: [] as string[], tail: '' };
    for (let i = 0; i < text.length; i += chunkSize) {
      last = splitter.advance(text.slice(0, i + chunkSize));
    }
    return { ...last, blocks: [...last.blocks] };
  }

  it('seals a block at a blank line and leaves the final segment in the tail', () => {
    const { blocks, tail } = feed('first\n\nsecond\n\nunfinished', 3);
    expect(blocks).toEqual(['first\n\n', 'second\n\n']);
    expect(tail).toBe('unfinished');
  });

  it('does not change a block once it has been emitted, which memoization depends on', () => {
    const splitter = new ReasoningBlockSplitter();
    splitter.advance('first\n\n');
    const firstSnapshot = splitter.advance('first\n\nseco')?.blocks[0];
    splitter.advance('first\n\nsecond\n\nthird');
    const stillSame = splitter.advance('first\n\nsecond\n\nthird more')?.blocks[0];
    expect(stillSame).toBe(firstSnapshot);
  });

  it('does not split on blank lines inside a code fence', () => {
    const text = 'intro\n\n```js\nconst a = 1;\n\nconst b = 2;\n```\n\noutro\n\n';
    const { blocks } = feed(text, 5);
    // The fence has to survive as one block and must not be split by the blank line inside it
    const fenceBlock = blocks.find((b) => b.includes('```'));
    expect(fenceBlock).toBeDefined();
    expect((fenceBlock!.match(/```/g) ?? []).length).toBe(2);
    expect(fenceBlock).toContain('const a = 1;');
    expect(fenceBlock).toContain('const b = 2;');
  });

  it('never seals an unclosed fence and keeps it in the tail until it closes', () => {
    const { blocks, tail } = feed('preface\n\n```js\nlet x = 1;\n\nlet y = 2;\n', 4);
    expect(blocks).toEqual(['preface\n\n']);
    expect(tail).toContain('```js');
    expect(tail).toContain('let y = 2;');
  });

  it('produces the same split regardless of chunk granularity', () => {
    const text = 'alpha\n\nbeta\n\n```py\nx=1\n\ny=2\n```\n\ngamma\n\ntailend';
    const baseline = feed(text, 1);
    for (const size of [2, 3, 7, 16, 64]) {
      const got = feed(text, size);
      expect(got.blocks).toEqual(baseline.blocks);
      expect(got.tail).toBe(baseline.tail);
    }
  });

  it('emits no empty blocks for consecutive blank lines, avoiding a swarm of empty components', () => {
    const { blocks } = feed('alpha\n\n\n\nbeta\n\n', 3);
    expect(blocks.every((b) => b.trim() !== '')).toBe(true);
  });

  it('rebuilds from scratch when the input is not an extension of the previous one (message switch or replay)', () => {
    const splitter = new ReasoningBlockSplitter();
    splitter.advance('old one\n\nold two\n\n');
    const { blocks, tail } = splitter.advance('brand new\n\nnew tail');
    expect(blocks).toEqual(['brand new\n\n']);
    expect(tail).toBe('new tail');
  });

  it('starts from zero after reset', () => {
    const splitter = new ReasoningBlockSplitter();
    splitter.advance('body\n\nmore');
    splitter.reset();
    const { blocks, tail } = splitter.advance('fresh\n\ntip');
    expect(blocks).toEqual(['fresh\n\n']);
    expect(tail).toBe('tip');
  });

  /**
   * Performance invariant: the tail is bounded. The tail is the only part reparsed each frame, so if it
   * grew with the total length, per-frame parse cost would rise with the thinking text and the total would
   * be back to O(n^2).
   */
  it('keeps the tail from growing with the total length of the thinking text', () => {
    const splitter = new ReasoningBlockSplitter();
    let text = '';
    let peakTail = 0;
    for (let i = 0; i < 2_000; i += 1) {
      text += `paragraph ${i} of reasoning, with **emphasis** and \`code\`.\n\n`;
      peakTail = Math.max(peakTail, splitter.advance(text).tail.length);
    }
    expect(peakTail).toBeLessThan(200);
  });

  /** The scan must be incremental: feeding the same text again must not emit the blocks twice. */
  it('is idempotent across repeated advance calls, which keeps it safe under StrictMode', () => {
    const splitter = new ReasoningBlockSplitter();
    const text = 'alpha\n\nbeta\n\ngamma';
    const first = splitter.advance(text);
    const second = splitter.advance(text);
    expect(second.blocks).toEqual(first.blocks);
    expect(second.tail).toBe(first.tail);
    expect(second.blocks.length).toBe(2);
  });
});

File: android/app/src/test/java/ai/oriveo/community/feature/chat/ReasoningStreamBlocksTest.kt
package ai.oriveo.community.feature.chat

import ai.oriveo.community.feature.chat.components.ReasoningBlockSplitter
import ai.oriveo.community.feature.chat.components.reasoningSafePrefix
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Behavior lock for the expanded reasoning stream's stable block splitting plus safe prefix.
 *
 * Two invariants:
 * - **Raw markers never reach the screen**: an unclosed `**` / `##` / table skeleton is held
 *   back until its shape is settled.
 * - **Cost is O(total input)**: once a block is sealed it never changes again (so `remember`
 *   can rely on it), and the tail does not grow with the total reasoning length (it is the
 *   only part re-parsed every frame; growing with the total would regress to O(n^2)).
 */
class ReasoningStreamBlocksTest {

    /** Feed input in chunks of a given size, simulating real chunk arrival */
    private fun feed(text: String, chunkSize: Int): Pair<List<String>, String> {
        val splitter = ReasoningBlockSplitter()
        var blocks: List<String> = emptyList()
        var tail = ""
        var i = 0
        while (i < text.length) {
            val end = minOf(i + chunkSize, text.length)
            val split = splitter.advance(text.substring(0, end))
            blocks = split.blocks
            tail = split.tail
            i = end
        }
        return blocks to tail
    }

    // MARK: - Block splitting

    @Test
    fun `blank line closes a block, last segment stays in tail`() {
        val (blocks, tail) = feed("first block\n\nsecond block\n\nunfinished", 3)
        assertEquals(listOf("first block\n\n", "second block\n\n"), blocks)
        assertEquals("unfinished", tail)
    }

    @Test
    fun `sealed block never changes afterwards`() {
        val splitter = ReasoningBlockSplitter()
        splitter.advance("first block\n\n")
        val first = splitter.advance("first block\n\nsecond").blocks[0]
        splitter.advance("first block\n\nsecond block\n\nthird block")
        val stillSame = splitter.advance("first block\n\nsecond block\n\nthird block more").blocks[0]
        assertEquals(first, stillSame)
    }

    @Test
    fun `blank line inside code fence does not split`() {
        val text = "intro\n\n```js\nconst a = 1;\n\nconst b = 2;\n```\n\noutro\n\n"
        val (blocks, _) = feed(text, 5)
        val fenceBlock = blocks.firstOrNull { it.contains("```") }
        assertTrue("fence block missing", fenceBlock != null)
        assertEquals(2, Regex("```").findAll(fenceBlock!!).count())
        assertTrue(fenceBlock.contains("const a = 1;"))
        assertTrue(fenceBlock.contains("const b = 2;"))
    }

    @Test
    fun `unclosed fence never seals, waits in tail`() {
        val (blocks, tail) = feed("preface\n\n```js\nlet x = 1;\n\nlet y = 2;\n", 4)
        assertEquals(listOf("preface\n\n"), blocks)
        assertTrue(tail.contains("```js"))
        assertTrue(tail.contains("let y = 2;"))
    }

    @Test
    fun `split result is independent of chunk size`() {
        val text = "one\n\ntwo\n\n```py\nx=1\n\ny=2\n```\n\nthree\n\ntail"
        val baseline = feed(text, 1)
        for (size in listOf(2, 3, 7, 16, 64)) {
            assertEquals("chunk=$size blocks differ", baseline.first, feed(text, size).first)
            assertEquals("chunk=$size tail differs", baseline.second, feed(text, size).second)
        }
    }

    @Test
    fun `non-extending input rebuilds from scratch`() {
        val splitter = ReasoningBlockSplitter()
        splitter.advance("old content\n\nold second block\n\n")
        val split = splitter.advance("brand new content\n\nnew tail")
        assertEquals(listOf("brand new content\n\n"), split.blocks)
        assertEquals("new tail", split.tail)
    }

    @Test
    fun `repeated advance is idempotent`() {
        val splitter = ReasoningBlockSplitter()
        val text = "one\n\ntwo\n\nthree"
        val first = splitter.advance(text)
        val second = splitter.advance(text)
        assertEquals(first.blocks, second.blocks)
        assertEquals(first.tail, second.tail)
        assertEquals(2, second.blocks.size)
    }

    /** The tail is the only part re-parsed every frame, so it must stay bounded. */
    @Test
    fun `tail does not grow with total reasoning length`() {
        val splitter = ReasoningBlockSplitter()
        val sb = StringBuilder()
        var peak = 0
        repeat(2_000) { i ->
            sb.append("Reasoning segment $i, with **emphasis** and `code`.\n\n")
            peak = maxOf(peak, splitter.advance(sb.toString()).tail.length)
        }
        assertTrue("tail peak $peak means content is accumulating, degraded to O(n^2)", peak < 200)
    }

    // MARK: - Safe prefix (no raw markers on screen)

    @Test
    fun `unclosed bold is held until closed`() {
        assertEquals("analysis", reasoningSafePrefix("analysis**key"))
        assertEquals("analysis**key constraint**, continue", reasoningSafePrefix("analysis**key constraint**, continue"))
    }

    @Test
    fun `no raw marker leaks at any prefix length`() {
        val source = "I need to **carefully** check: call `flush()` and then continue."
        for (len in 1..source.length) {
            val shown = reasoningSafePrefix(source.substring(0, len))
            // Paired, closed markers are allowed to remain in the source text (markdown rendering
            // handles them), but an unclosed opening marker must never leak -- the check: `**`
            // must appear an even number of times
            assertEquals(
                "len=$len leaked unclosed **: $shown",
                0,
                Regex("\\*\\*").findAll(shown).count() % 2,
            )
            assertEquals(
                "len=$len leaked unclosed backtick: $shown",
                0,
                shown.count { it == '`' } % 2,
            )
        }
    }

    @Test
    fun `heading ambiguity window holds`() {
        // Just `##` with no space yet -- shape undetermined, so hold it back; otherwise it would
        // render as body text before flipping to a heading
        assertEquals("", reasoningSafePrefix("##"))
        assertEquals("## Step two", reasoningSafePrefix("## Step two"))
    }

    @Test
    fun `hr candidate holds until line completes`() {
        assertEquals("", reasoningSafePrefix("---"))
        assertEquals("preceding text\n", reasoningSafePrefix("preceding text\n--"))
    }

    @Test
    fun `table skeleton is held until region completes`() {
        // The header row can't be confirmed as a table yet (depends on whether the next line is
        // a separator) -- hold it back
        assertFalse(reasoningSafePrefix("text\n| ColA | ColB |").contains("|"))
        assertFalse(reasoningSafePrefix("text\n| ColA | ColB |\n|---|---|\n").contains("|---|"))
    }

    @Test
    fun `unclosed block math is held`() {
        assertFalse(reasoningSafePrefix("text\n$$\nx = 1\n").contains("$$"))
    }

    @Test
    fun `plain text passes through untouched`() {
        assertEquals("Plain reasoning content, no markers at all.", reasoningSafePrefix("Plain reasoning content, no markers at all."))
        assertEquals("first line\nsecond line", reasoningSafePrefix("first line\nsecond line"))
    }

    @Test
    fun `empty input yields empty`() {
        assertEquals("", reasoningSafePrefix(""))
    }
}
